File: backend/src/main/java/com/example/demo/config/RedisConfig.java
package com.example.demo.config;

import com.example.demo.service.RedisSubscriber;
import lombok.RequiredArgsConstructor;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.data.redis.connection.RedisConnectionFactory;
import org.springframework.data.redis.listener.ChannelTopic;
import org.springframework.data.redis.listener.RedisMessageListenerContainer;
import org.springframework.data.redis.listener.adapter.MessageListenerAdapter;

@Configuration
@RequiredArgsConstructor
public class RedisConfig {

    @Bean
    public ChannelTopic topic() {
        return new ChannelTopic("notification-topic");
    }

    @Bean
    public MessageListenerAdapter messageListener(RedisSubscriber subscriber) {
        return new MessageListenerAdapter(subscriber);
    }

    @Bean
    public RedisMessageListenerContainer redisMessageListener(
            RedisConnectionFactory connectionFactory,
            MessageListenerAdapter listenerAdapter,
            ChannelTopic topic) {

        RedisMessageListenerContainer container = new RedisMessageListenerContainer();
        container.setConnectionFactory(connectionFactory);
        container.addMessageListener(listenerAdapter, topic);
        return container;
    }
}

File: backend/src/main/kotlin/com/example/demo/config/RedisConfig.kt
package com.example.demo.config

import com.example.demo.service.RedisSubscriber
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.data.redis.connection.RedisConnectionFactory
import org.springframework.data.redis.listener.ChannelTopic
import org.springframework.data.redis.listener.RedisMessageListenerContainer
import org.springframework.data.redis.listener.adapter.MessageListenerAdapter

@Configuration
class RedisConfig {

    @Bean("notificationTopic")
    fun notificationTopic(): ChannelTopic {
        return ChannelTopic("notification-topic")
    }

    @Bean("notificationMessageListener")
    fun notificationMessageListener(subscriber: RedisSubscriber): MessageListenerAdapter {
        return MessageListenerAdapter(subscriber)
    }

    @Bean("notificationMessageListenerContainer")
    fun notificationMessageListenerContainer(
            connectionFactory: RedisConnectionFactory,
            notificationMessageListener: MessageListenerAdapter,
            notificationTopic: ChannelTopic
    ): RedisMessageListenerContainer {
        val container = RedisMessageListenerContainer()
        container.setConnectionFactory(connectionFactory)
        container.addMessageListener(notificationMessageListener, notificationTopic)
        return container
    }

}
